### PR TITLE
#1crkd2n Round Rotation

### PR DESCRIPTION
**Pull request by BrayonCrayon**
- Implemented round rotation
- Removed user connection to black cards
- Add judge to game for better access to user table

### CODE REVIEW

File: app/Http/Controllers/Game/RotateGameController.php [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+        $newBlackCard = BlackCard::query()
+            ->inRandomOrder()
+            ->whereNotIn('id', $game->gameBlackCards->pluck('black_card_id'))
+            ->first();
+
+        GameBlackCards::create([
+            'game_id' => $game->id,
+            'black_card_id' => $newBlackCard->id
+        ]);
```
Comment (BrayonCrayon) [inline app/Http/Controllers/Game/RotateGameController.php:45]: Could be used in Game service

File: app/Http/Controllers/Game/RotateGameController.php [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+            if ($pair->first()->id === $game->judge_id) {
+                $game->update([
+                    'judge_id' => $pair->last()->id
+                ]);
+                return false;
+            }
+            return true;
+        });
+
+        $game->gameBlackCards()->first()->delete();
```
Comment (BrayonCrayon) [inline app/Http/Controllers/Game/RotateGameController.php:35]: Put into GameService

File: app/Http/Controllers/Game/RotateGameController.php [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+        $users->sliding(2)->each(function($pair) use ($game) {
+            if ($pair->first()->id === $game->judge_id) {
+                $game->update([
+                    'judge_id' => $pair->last()->id
+                ]);
+                return false;
+            }
+            return true;
+        });
```
Comment (BrayonCrayon) [inline app/Http/Controllers/Game/RotateGameController.php:33]: Put into GameService

File: tests/Feature/Http/Controllers/Game/RoundRotationTest.php [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+        $blackCardPick = $this->game->gameBlackCards()->first()->blackCard->pick;
+
+        $pickedJudgeIds = collect();
+
+        $this->game->users->each(function ($user) use ($blackCardPick, $pickedJudgeIds) {
+
+            $this->game->users->each(function($user) use($blackCardPick) {
+                $userCards = $user->whiteCardsInGame->take($blackCardPick);
+                $userCards->each(fn ($card) => $card->update(['selected' => true]));
+            });
```
Comment (BrayonCrayon) [inline tests/Feature/Http/Controllers/Game/RoundRotationTest.php:66]: Extract to function in base test class

File: tests/Feature/Http/Controllers/Game/RoundRotationTest.php [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+        $users = User::factory(5)->create();
+        $this->game = Game::factory()->create([
+            'judge_id' => $users->first()->id,
+        ]);
+        foreach ($users as $user) {
+            $gameService->grabWhiteCards($user, $this->game, $this->expansionIds);
+            $this->game->users()->save($user);
+        }
+        $this->game->expansions()->saveMany(Expansion::idsIn($this->expansionIds)->get());
+        $gameService->grabBlackCards($users->first(), $this->game, $this->expansionIds);
```
Comment (BrayonCrayon) [inline tests/Feature/Http/Controllers/Game/RoundRotationTest.php:34]: Use factories to setup

File: tests/Feature/Http/Controllers/Game/RoundRotationTest.php [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+
+        $pickedJudgeIds = collect();
+
+        $this->game->refresh();
+        $this->game->users->each(function ($user) use ($blackCardPick, $pickedJudgeIds) {
+
+            $this->game->users->each(function($user) use($blackCardPick) {
+                $userCards = $user->whiteCardsInGame->take($blackCardPick);
+                $userCards->each(fn ($card) => $card->update(['selected' => true]));
+            });
```
Comment (BrayonCrayon) [inline tests/Feature/Http/Controllers/Game/RoundRotationTest.php:96]: Here too

File: tests/Feature/Http/Controllers/Game/RoundRotationTest.php [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+        $this->postJson(route('api.game.rotate', $this->game->id))->assertOk();
+
+        $this->game->refresh();
+        $this->assertNotEquals($firstJudge->id, $this->game->judge_id);
+    }
+
+    /** @test */
+    public function it_cycles_through_the_users_when_assigning_the_judge_when_number_of_users_is_odd()
+    {
+        $blackCardPick = $this->game->gameBlackCards()->first()->blackCard->pick;
```
Comment (BrayonCrayon) [inline tests/Feature/Http/Controllers/Game/RoundRotationTest.php:57]: Use current black card attribute off of game model

File: tests/Feature/Http/Controllers/Game/RoundRotationTest.php [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+    {
+        /** @var Game $game */
+        $game = $this->game;
+        $blackCardPick = $game->gameBlackCards()->first()->blackCard->pick;
+        $previousBlackCard = $game->currentBlackCard;
+
+        $game->users->each(function($user) use($blackCardPick) {
+            $userCards = $user->whiteCardsInGame->take($blackCardPick);
+            $userCards->each(fn ($card) => $card->update(['selected' => true]));
+        });
```
Comment (BrayonCrayon) [inline tests/Feature/Http/Controllers/Game/RoundRotationTest.php:122]: Same extraction method

**Review by ianvanharten — APPROVED** [review]

**Pull request merged**